core/src/data_sources/splitter.rs

```diff
@@ -103,13 +105,86 @@ impl Splitter for BaseV0Splitter {
         if chunk.len() > 0 {
             chunks.push(chunk);
         }
+        let mut decoded = Vec::new();
+        let mut remaining : Vec<usize> = Vec::new();
+
+        fn decode_chunk_with_remainder(embedder: &Box<dyn Embedder + Sync + Send>, chunk: Vec<usize>) -> Result<(String, Vec<usize>)> {
+            let mut result : String = String::new();
+            let mut remaining : Vec<usize> = Vec::new();
+
+            let mut end = chunk.len();
+            
+            while end > 0 {
+                let decoded_future = embedder.decode(chunk.clone()[0..end].to_vec());
```

Comment:
`decode(...).await?`

Will await the future and also return the value directly with `?` (no need to unwrap). If an error occurred it will make the function return with that error.

If you want to custom handle the case where there is an error, use a `match` syntax here:

```
match embeder.decode(...).await {
  Ok(x) => ...,
  Err(e) => ...,
}
```